nova/scheduler/filters/external_scheduler_filter.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2024 SAP SE or an SAP affiliate company.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+import requests
+
+import nova.conf
+from nova.scheduler import filters
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+CONF = nova.conf.CONF
+
+
+class ExternalSchedulerFilter(filters.BaseHostFilter):
+    """Filter using an out of tree scheduler via a REST API."""
+
+    RUN_ON_REBUILD = False
+
+    def __init__(self):
+        super().__init__()
+        self.api_url = CONF.filter_scheduler.external_scheduler_api_url
+
+    def filter_all(self, filter_obj_list, spec_obj):
+        """Yield objects that pass the filter."""
+        if not self.api_url:
+            LOG.debug(
+                "external_scheduler_api_url not configured. skipping filter"
+            )
+            return filter_obj_list
+
+        hosts_data = []
+        for obj in filter_obj_list:
+            hosts_data.append({
+                "name": obj.host,
+                "status": obj.status,
+                "vcpus_total": obj.vcpus_total,
+                "vcpus_used": obj.vcpus_used,
+                "memory_mb": obj.memory_mb,
+                "memory_mb_used": obj.memory_mb_used
+            })
+
+        spec_data = {
+            "vcpus": spec_obj.vcpus,
+            "memory_mb": spec_obj.memory_mb,
+        }
+
+        json_data = {
+            "hosts": hosts_data,
+            "request_spec": spec_data
+        }
+
+        try:
+            response = requests.post(
+                self.api_url, json=json_data, timeout=10
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            valid_hosts = response_json.get('valid_hosts', [])
+            for obj in filter_obj_list:
+                if obj.host in valid_hosts:
+                    yield obj
+
+        except requests.RequestException as e:
+            # Log an error if the request fails
```

Comment:
I think this comment doesn't add anything helpful as the next line starts with `LOG.error` and thus easily conveys the same information.